storagenode/retain/retain.go

```diff
@@ -88,105 +89,207 @@ type Service struct {
 	log    *zap.Logger
 	config Config
 
-	mu     sync.Mutex
-	queued map[storj.NodeID]Request
+	cond    sync.Cond
+	queued  map[storj.NodeID]Request
+	working map[storj.NodeID]struct{}
+	group   errgroup.Group
 
-	reqChan      chan Request
-	semaphore    chan struct{}
-	emptyTrigger chan struct{}
+	closedOnce sync.Once
+	closed     chan struct{}
+	started    bool
 
 	store *pieces.Store
 }
 
 // NewService creates a new retain service.
 func NewService(log *zap.Logger, store *pieces.Store, config Config) *Service {
 	return &Service{
-		log:          log,
-		config:       config,
-		queued:       make(map[storj.NodeID]Request),
-		reqChan:      make(chan Request),
-		semaphore:    make(chan struct{}, config.MaxConcurrentRetain),
-		emptyTrigger: make(chan struct{}),
-		store:        store,
+		log:    log,
+		config: config,
+
+		cond:    *sync.NewCond(&sync.Mutex{}),
+		queued:  make(map[storj.NodeID]Request),
+		working: make(map[storj.NodeID]struct{}),
+		closed:  make(chan struct{}),
+
+		store: store,
 	}
 }
 
 // Queue adds a retain request to the queue.
 // It discards a request for a satellite that already has a queued request.
 // true is returned if the request is queued and false is returned if it is discarded
 func (s *Service) Queue(req Request) bool {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.cond.L.Lock()
+	defer s.cond.L.Unlock()
 
-	// subtract some time to leave room for clock difference between the satellite and storage node
-	req.CreatedBefore = req.CreatedBefore.Add(-s.config.RetainTimeBuffer)
+	select {
+	case <-s.closed:
+		return false
+	default:
+	}
 
-	// only queue retain request if we do not already have one for this satellite
-	if _, ok := s.queued[req.SatelliteID]; !ok {
-		s.queued[req.SatelliteID] = req
-		go func() { s.reqChan <- req }()
+	// subtract some time to leave room for clock difference between the satellite and storage node
+	req.CreatedBefore = req.CreatedBefore.Add(-s.config.MaxTimeSkew)
 
-		return true
-	}
+	s.queued[req.SatelliteID] = req
+	s.cond.Broadcast()
```

Comment:
Is there a reason it's now waking up all workers?

Reply:
Personally, I find that `Signal` is much harder to reason about and harder to maintain correctness of. It depends on the fact that ANYONE in a `(*Cond).Wait` call is available to handle the change in state, but that's not always the case. 

I started this comment by saying "Hmm, I guess not.", but then I realized that if there's a pending `(*Service).Wait()` call, and `Signal` only wakes up that one, no worker will wake up to handle the queued request, leading to deadlock.

Reply:
Yeah, that's why used polling, since `Wait` is only used for testing... and `Signal` can avoid spurious wake ups. But, I do agree that it's harder to reason about. Since this isn't called too often, I'm also fine with Broadcast.